rust/src/enip/detect.rs

```diff
@@ -0,0 +1,391 @@
+/* Copyright (C) 2023 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use nom7::branch::alt;
+use nom7::character::complete::{char, digit1, space0};
+use nom7::combinator::{map_opt, opt, verify};
+use nom7::error::{make_error, ErrorKind};
+use nom7::IResult;
+
+use std::ffi::c_void;
+
+use crate::enip::enip::EnipTransaction;
+use crate::enip::parser::{
+    CipData, CipDir, EnipCipRequestPayload, EnipCipResponsePayload, EnipItemPayload, EnipPayload,
+    CIP_MULTIPLE_SERVICE, ENIP_CIP_PATH_ATTR_16BIT, ENIP_CIP_PATH_ATTR_8BIT,
+    ENIP_CIP_PATH_CLASS_16BIT, ENIP_CIP_PATH_CLASS_8BIT, ENIP_CMD_CANCEL, ENIP_CMD_INDICATE_STATUS,
+    ENIP_CMD_LIST_IDENTITY, ENIP_CMD_LIST_INTERFACES, ENIP_CMD_LIST_SERVICES, ENIP_CMD_NOP,
+    ENIP_CMD_REGISTER_SESSION, ENIP_CMD_SEND_RRDATA, ENIP_CMD_SEND_UNIT_DATA,
+    ENIP_CMD_UNREGISTER_SESSION,
+};
+
+use crate::core::Direction;
+
+use std::ffi::CStr;
+
+fn enip_detect_parse_u16(i: &str) -> IResult<&str, u16> {
+    let (i, r) = map_opt(digit1, |s: &str| s.parse::<u16>().ok())(i)?;
+    return Ok((i, r));
+}
+
+fn enip_parse_command_string(i: &str) -> IResult<&str, u16> {
+    let su = i.to_uppercase();
+    let su_slice: &str = &su;
+    match su_slice {
+        "NOP" => Ok((i, ENIP_CMD_NOP)),
+        "LISTSERVICES" => Ok((i, ENIP_CMD_LIST_SERVICES)),
+        "LISTIDENTITY" => Ok((i, ENIP_CMD_LIST_IDENTITY)),
+        "LISTINTERFACES" => Ok((i, ENIP_CMD_LIST_INTERFACES)),
+        "REGISTERSESSION" => Ok((i, ENIP_CMD_REGISTER_SESSION)),
+        "UNREGISTERSESSION" => Ok((i, ENIP_CMD_UNREGISTER_SESSION)),
+        "SENDRRDATA" => Ok((i, ENIP_CMD_SEND_RRDATA)),
+        "SENDUNITDATA" => Ok((i, ENIP_CMD_SEND_UNIT_DATA)),
+        "INDICATESTATUS" => Ok((i, ENIP_CMD_INDICATE_STATUS)),
+        "CANCEL" => Ok((i, ENIP_CMD_CANCEL)),
+        _ => Err(nom7::Err::Error(nom7::error::make_error(
+            i,
+            nom7::error::ErrorKind::MapOpt,
+        ))),
+    }
+}
+
+fn enip_parse_command(i: &str) -> IResult<&str, u16> {
+    let (i, v) = alt((enip_detect_parse_u16, enip_parse_command_string))(i)?;
+    return Ok((i, v));
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_enip_parse_command(
```

Comment:
Public C API, so `SCEnipParseCommand`?

Reply:
Basically apply this to any no_mange pub extern C fn.

Reply:
We should probably have a (tracking) ticket to clean this up for any existing code. Plus see if we can have a CI check?